### ISSUE TITLE
Dict => JSON => Dict round trip fails in the Python dataframe API

### ISSUE THREAD

**Issue by ixxie** [duckdb/duckdb#4303]
### What happens?

When inserting Python dictionaries from a dataframe into a JSON column, and then querying them back into a dataframe, the type is cast into a Python dictionary rendered as a string. This forces the user to resort to unsecure methods such as `eval` in order to easily reload the data into Python.

### To Reproduce

```python
import duckdb
import pandas as pd

con = duckdb.connect(':memory:')

con.execute('create table test (j JSON);')

test_df = pd.DataFrame.from_dict({'j': [{'a':1}, {'b': 'foo'}, {'c': True}]})
print(test_df)
#                j
#  0      {'a': 1}
#  1  {'b': 'foo'}
#  2   {'c': True

con.execute('insert into test (select * from test_df);')

result = con.execute('select * from test;').fetchdf().to_dict('records')
print(result)
#  [{'j': "{'a': 1}"}, {'j': "{'b': 'foo'}"}, {'j': "{'c': True}"}]
```

### OS:

NixOS Unstable

### DuckDB Version:

0.4.0

### DuckDB Client:

Python

### Full Name:

Matan Shenhav

### Affiliation:

Karsa Oy

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Tishj** [comment]
Hi, thank you for raising the issue!

I am however confused by the answers you filled out, as the DuckDB version says 0.4.0.
Have you verified that this issue still exists on the latest master branch?

**Comment by ixxie** [comment]
I have not verified, because I am not entirely sure how to do this in Poetry... I have a bit of an unusual setup and I am not exactly sure how to run such a test.

**Comment by Mause** [comment]
> I have not verified, because I am not entirely sure how to do this in Poetry... I have a bit of an unusual setup and I am not exactly sure how to run such a test.

You can run `pip install duckdb -U --pre` inside the poetry venv to try it out

**Comment by ixxie** [comment]
I'm swamped at the moment and will go on holiday soon; I will have to look into this in September...

**Comment by lnkuiper** [comment]
I don't think we support this yet, even in the latest master. To insert python dicts into a JSON column, we'd need some kind of conversion within the python client.

Right now, we already have a conversion from python dicts to our `MAP` type, which can be cast to JSON, but I don't think this work the way you'd want it to work, yet.

**Comment by lnkuiper** [comment]
Coming back to this, I think this is solved by using `to_json`, i.e.:
```python3
import duckdb
import pandas as pd

con = duckdb.connect(':memory:')

con.execute('create table test (j JSON);')

test_df = pd.DataFrame.from_dict({'j': [{'a':1}, {'b': 'foo'}, {'c': True}]})
print(test_df)
#                j
#  0      {'a': 1}
#  1  {'b': 'foo'}
#  2   {'c': True

con.execute('insert into test (select to_json(j) from test_df);')

result = con.execute('select * from test;').fetchdf()
print(result)
#               j
# 0     {"a":"1"}
# 1   {"b":"foo"}
# 2  {"c":"true"}
```
However, I'm not sure if this is what you want:
```python3
print(result.to_dict('records'))
# [{'j': '{"a":"1"}'}, {'j': '{"b":"foo"}'}, {'j': '{"c":"true"}'}]
```

**Comment by lnkuiper** [comment]
Just wanted to add this: Inserting any kind of data into a column of type `JSON` will cast it to `VARCHAR`. I could add special cases for, `LIST`, `MAP` and `STRUCT`, to automatically use `to_json` instead. This would make it more intuitive to insert things into a `JSON` column.

However, to get the data out of a `JSON` column into Python would require our Python client to be aware of the `JSON` type, rather than just treating it as a string.

**Comment by lnkuiper** [comment]
Closed via #7366
